util/circular_buffer.c

```diff
@@ -141,105 +147,195 @@ int32_t cb_remove(struct circular_buffer *desc)
 /**
  * @brief Get the number of elements in the buffer
  * @param desc - Circular buffer reference
- * @param nb_elements - Where to store the number of elements
+ * @param size - Where to store size of data available to read
  * @return
- *  - \ref SUCCESS : On success
- *  - \ref FAILURE : Otherwise
+ *  - \ref SUCCESS   - No errors
+ *  - -EINVAL   - Wrong parameters used
+ *  - -EOVERRUN - A buffer overrun occurred
  */
-int32_t cb_size(struct circular_buffer *desc, uint32_t *nb_elements)
+int32_t cb_size(struct circular_buffer *desc, uint32_t *size)
 {
-	if (!desc)
-		return FAILURE;
+	uint32_t nb_spins;
 
-	*nb_elements =  desc->nb_elem;
+	if (!desc || !size)
+		return -EINVAL;
+
+	if (desc->write.spin_count > desc->read.spin_count)
+		nb_spins = desc->write.spin_count - desc->read.spin_count;
+	else
+		/* Integer overflow on desc->write.spin_count */
+		nb_spins = UINT32_MAX - desc->read.spin_count +
+			   desc->write.spin_count + 1;
+
+	if (nb_spins > 0)
+		*size = desc->size + desc->write.idx - desc->read.idx;
+	else
+		*size = desc->write.idx - desc->read.idx;
+
+	if (*size > desc->size) {
+		*size = desc->size;
+		return -EOVERRUN;
```

Comment:
I don't get how this is possible.
Maybe the term is wrong ? Overrun is reading past a buffer size. This happens in hardware FIFO (think of an SPI FIFO) but I don't understand how it's possible to have this error in a circular buffer implementation where you control how much the user reads. You can block reads of larger size than what is available in the circular buffer.

Reply:
Overrun on a circular buffer occurs when you write more data than the space available and the old data is overwritten.
So, this error is returned on a read operation in order to know that some data have been overwritten. 
In the app then, you can ignore this error if old data is useless (ex audio buffers) or you know that you should use a bigger buffer.